pkg/controller/networkpolicy/networkpolicy_controller.go

```diff
@@ -1061,6 +1081,67 @@ func (n *NetworkPolicyController) deleteNamespace(old interface{}) {
 	}
 }
 
+// addService retrieves all internal Groups which refers to this Service
+// and enqueues the group keys for further processing.
+func (n *NetworkPolicyController) addService(obj interface{}) {
+	defer n.heartbeat("addService")
+	service := obj.(*v1.Service)
+	klog.V(2).Infof("Processing Service %s/%s ADD event", service.Namespace, service.Name)
+	// Find all internal Group keys which refers to this Service.
+	groupKeySet := n.filterInternalGroupsForService(service)
+	// Enqueue internal groups to its queue for group processing.
+	for group := range groupKeySet {
+		n.enqueueInternalGroup(group)
+	}
+}
+
+// updatePod retrieves all internal Groups which refers to this Service
+// and enqueues the group keys for further processing.
+func (n *NetworkPolicyController) updateService(oldObj, curObj interface{}) {
+	defer n.heartbeat("updateService")
+	oldService := oldObj.(*v1.Service)
+	curService := curObj.(*v1.Service)
+	klog.V(2).Infof("Processing Service %s/%s UPDATE event, selectors: %v", curService.Namespace, curService.Name, curService.Spec.Selector)
+	// No need to trigger processing of groups if there is no change in the Service selectors.
+	if reflect.DeepEqual(oldService.Spec.Selector, curService.Spec.Selector) {
+		klog.V(4).Infof("No change in Service %s/%s. Skipping group evaluation.", curService.Namespace, curService.Name)
+		return
+	}
+	// Find all internal Group keys which refers to this Service.
+	groupKeySet := n.filterInternalGroupsForService(curService)
```

Comment:
Dont we need to handle old groups?

Reply:
Can you elaborate on what 'old groups' mean here? If a CG is created with reference to Service, then a Service CRUD will not affect this association. We merely need to update the computed selectors for this CG. On the other hand, if a Service CG is update/deleted, it'll be handled by CG sync loops, not here.

Reply:
You are right. I was thinking wrongly that services are selected by labels.